[BUG] Document Link custom field "contains" filter crashes with TypeError when a linked document has another unset Document Link field

### Description

Filtering documents by a Document Link custom field using the contains operator crashes with an unhandled 500 Internal Server Error if the document referenced in the query has any other Document Link custom field attached to it that has never been given a value.

<img width="1565" height="986" alt="Image" src="https://github.com/user-attachments/assets/19454d95-2044-463c-ab0f-b8ed0aa7897a" />
(debug mode enabled)

### Expected behavior

The filter returns the document(s) whose related field actually contains the target document, even if unset custom fields exist.

### Actual behavior

The request fails with an unhandled server error. Error while loading documents: TypeError:
```
  File ".../documents/filters.py", line 740, in _parse_atom_doc_link_contains
    document_ids_we_want = functools.reduce(
  File ".../documents/filters.py", line 742, in <genexpr>
    (set(link.value_document_ids) for link in links),
  TypeError: 'NoneType' object is not iterable
```

### Steps to reproduce

- Create 2 simple documents: Document A and Document B
- Create a Document Link custom field, e.g. named `related`.
- Open Document B's detail view → Custom Fields → add `related` → select Document A as the target.
- Add a second, unrelated Document Link custom field to Document B (or Document A), e.g. named `duplicate`, via the Custom Fields dropdown, but do not select any target document for it (leave it unset).
- On the documents list, open the "Custom fields" query builder → build: related contains Document B (or Document A, whichever now carries the unset duplicate field).
- Apply the filter.


### Webserver logs

```bash
[03/Aug/2026 07:58:44] "GET /api/documents/?page=1&ordering=-created&truncate_content=true HTTP/1.1" 200 1245
Internal Server Error: /secretpath/api/documents/
Traceback (most recent call last):
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django/core/handlers/except
ion.py", line 55, in inner
    response = get_response(request)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django/core/handlers/base.p
y", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django/views/decorators/csr
f.py", line 65, in _view_wrapper
    return view_func(request, *args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/viewsets.py"
, line 125, in view
    return self.dispatch(request, *args, **kwargs)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/views.py", l
ine 515, in dispatch
    response = self.handle_exception(exc)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/views.py", l
ine 475, in handle_exception
    self.raise_uncaught_exception(exc)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/views.py", l
ine 486, in raise_uncaught_exception
    raise exc
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/views.py", l
ine 512, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/views.py", line 2365, in list
    return super().list(request)
           ~~~~~~~~~~~~^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/views.py", line 1251, in list
    queryset = self.filter_queryset(self.get_queryset())
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/generics.py"
, line 154, in filter_queryset
    queryset = backend().filter_queryset(self.request, queryset, self)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django_filters/rest_framewo
rk/backends.py", line 72, in filter_queryset
    return filterset.qs
           ^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django_filters/filterset.py
", line 250, in qs
    qs = self.filter_queryset(qs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django_filters/filterset.py
", line 233, in filter_queryset
    queryset = self.filters[name].filter(queryset, value)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 780, in filter
    q, annotations = parser.parse(value)
                     ~~~~~~~~~~~~^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 475, in parse
    self._parse_expr(expr, validation_prefix=self._validation_prefix),
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 360, in wrapper
    return func(*args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 487, in _parse_expr
    return self._parse_logical_expr(*expr)
           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 525, in _parse_logical_expr
    qs = self._parse_expr_list(args, validation_prefix="1")
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 360, in wrapper
    return func(*args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 504, in _parse_expr_list
    self._parse_expr(expr, validation_prefix=i) for i, expr in enumerate(exprs)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 360, in wrapper
    return func(*args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 489, in _parse_expr
    return self._parse_atom(*expr)
           ~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 553, in _parse_atom
    return self._parse_atom_doc_link_contains(custom_field, value)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 740, in _parse_atom_doc_link_c
ontains
    document_ids_we_want = functools.reduce(
        operator.and_,
        (set(link.value_document_ids) for link in links),
    )
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 742, in <genexpr>
    (set(link.value_document_ids) for link in links),
     ~~~^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not iterable
[2026-08-03 07:58:44,464] [ERROR] [django.request] Internal Server Error: /secretpath/api/documents/
Traceback (most recent call last):
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django/core/handlers/except
ion.py", line 55, in inner
    response = get_response(request)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django/core/handlers/base.p
y", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django/views/decorators/csr
f.py", line 65, in _view_wrapper
    return view_func(request, *args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/viewsets.py"
, line 125, in view
    return self.dispatch(request, *args, **kwargs)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/views.py", l
ine 515, in dispatch
    response = self.handle_exception(exc)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/views.py", l
ine 475, in handle_exception
    self.raise_uncaught_exception(exc)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/views.py", l
ine 486, in raise_uncaught_exception
    raise exc
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/views.py", l
ine 512, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/views.py", line 2365, in list
    return super().list(request)
           ~~~~~~~~~~~~^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/views.py", line 1251, in list
    queryset = self.filter_queryset(self.get_queryset())
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/rest_framework/generics.py"
, line 154, in filter_queryset
    queryset = backend().filter_queryset(self.request, queryset, self)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django_filters/rest_framewo
rk/backends.py", line 72, in filter_queryset
    return filterset.qs
           ^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django_filters/filterset.py
", line 250, in qs
    qs = self.filter_queryset(qs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/.venv/lib/python3.14/site-packages/django_filters/filterset.py
", line 233, in filter_queryset
    queryset = self.filters[name].filter(queryset, value)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 780, in filter
    q, annotations = parser.parse(value)
                     ~~~~~~~~~~~~^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 475, in parse
    self._parse_expr(expr, validation_prefix=self._validation_prefix),
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 360, in wrapper
    return func(*args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 487, in _parse_expr
    return self._parse_logical_expr(*expr)
           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 525, in _parse_logical_expr
    qs = self._parse_expr_list(args, validation_prefix="1")
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 360, in wrapper
    return func(*args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 504, in _parse_expr_list
    self._parse_expr(expr, validation_prefix=i) for i, expr in enumerate(exprs)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 360, in wrapper
    return func(*args, **kwargs)
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 489, in _parse_expr
    return self._parse_atom(*expr)
           ~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 553, in _parse_atom
    return self._parse_atom_doc_link_contains(custom_field, value)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 740, in _parse_atom_doc_link_c
ontains
    document_ids_we_want = functools.reduce(
        operator.and_,
        (set(link.value_document_ids) for link in links),
    )
  File "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/src/documents/filters.py", line 742, in <genexpr>
    (set(link.value_document_ids) for link in links),
     ~~~^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not iterable
[03/Aug/2026 07:58:44] "GET /api/documents/?page=1&page_size=50&ordering=-created&truncate_content=true&include_selec
tion_data=true&custom_field_query=%5B%22OR%22,%5B%5B9,%22contains%22,%5B1%5D%5D%5D%5D HTTP/1.1" 500 10829
```

### Browser logs

```bash
N/A (same HTTP 500)
```

### Services logs

```bash
N/A
```

### Paperless-ngx version

3.0.5

### Host OS

Ubuntu 26.04

### Installation method

Bare metal

### System status

```json
{
    "pngx_version": "3.0.5",
    "server_os": "Linux-7.0.0-28-generic-x86_64-with-glibc2.43",
    "install_type": "bare-metal",
    "storage": {
        "total": 2011859927040,
        "available": 1165483646976
    },
    "database": {
        "type": "sqlite",
        "url": "/home/ggouzi/Documents/DEV/Perso/paperless-ngx/data/db.sqlite3",
        "status": "OK",
        "error": null,
        "migration_status": {
            "latest_migration": "paperless_mail.0004_processedmail_uid_validity",
            "unapplied_migrations": []
        }
    },
    "tasks": {
        "redis_url": "redis://localhost:6379",
        "redis_status": "OK",
        "redis_error": null,
        "celery_status": "OK",
        "celery_url": "celery@ggouzi-ThinkPad-P16",
        "celery_error": null,
        "index_status": "OK",
        "index_last_modified": "2026-08-03T07:44:16.832897Z",
        "index_error": null,
        "classifier_status": "WARNING",
        "classifier_last_trained": null,
        "classifier_error": "No classifier training tasks found",
        "sanity_check_status": "WARNING",
        "sanity_check_last_run": null,
        "sanity_check_error": "No sanity check tasks found",
        "llmindex_status": "DISABLED",
        "llmindex_last_modified": null,
        "llmindex_error": null,
        "summary": {
            "days": 30,
            "total_count": 2,
            "pending_count": 0,
            "success_count": 1,
            "failure_count": 1
        }
    },
    "websocket_connected": "ERROR"
}
```

### Browser

Firefox

### Configuration changes
```
PAPERLESS_SECRET_KEY=xI0rlgMZfUIb9fjQ_PmXNjrPeh7Y_esO5OENKXwZK7aCPY_E8UvNqa6qQ0TD-cI2uvxFleCOq22Jj7MYksdOLg
PAPERLESS_DEBUG=true
PAPERLESS_CORS_ALLOWED_HOSTS=http://localhost:4200,http://192.168.1.4:4200
PAPERLESS_CSRF_TRUSTED_ORIGINS=http://localhost:4200,http://192.168.1.4:4200
PAPERLESS_FORCE_SCRIPT_NAME=/secretpath
PAPERLESS_OCR_DESKEW=true
PAPERLESS_OCR_MODE=auto
```
### Please confirm the following

- [x] I believe this issue is a bug that affects all users of Paperless-ngx, not something specific to my installation.
- [x] This issue is not about the OCR or archive creation of a specific file(s). Otherwise, please see above regarding OCR tools.
- [x] I have already searched for relevant existing issues and discussions before opening this report.
- [x] I have updated the title field above with a concise description.

--- a/src/documents/filters.py
+++ b/src/documents/filters.py
@@ -725,9 +725,13 @@ class CustomFieldQueryParser:
             )
 
         # First we look up reverse links from the requested documents.
+        # Scoped to this specific field (not just any document link field) and
+        # excluding unset instances, which have a null value_document_ids and
+        # are equivalent to having no reverse link at all.
         links = CustomFieldInstance.objects.filter(
             document_id__in=value,
-            field__data_type=CustomField.FieldDataType.DOCUMENTLINK,
+            field=custom_field,
+            value_document_ids__isnull=False,
         )
 
         # Check if any of the requested IDs are missing.

--- a/src/documents/tests/test_api_filter_by_custom_fields.py
+++ b/src/documents/tests/test_api_filter_by_custom_fields.py
@@ -8,6 +8,7 @@ from django.contrib.auth.models import User
 from rest_framework.test import APITestCase
 
 from documents.models import CustomField
+from documents.models import CustomFieldInstance
 from documents.models import Document
 from documents.models import SavedView
 from documents.models import SavedViewFilterRule
@@ -604,6 +605,56 @@ class TestCustomFieldsSearch(DirectoriesMixin, APITestCase):
                 and set(document["documentlink_field"]) >= {self.documents[6].id}
             ),
             match_nothing_ok=True,
+        )
+
+    def test_document_link_contains_unset_reverse_link(self) -> None:
+        # Another edge case: the document in the value list has the same
+        # document link field attached, but it was never given a value
+        # (value_document_ids is None). This must be treated the same as
+        # having no reverse link at all, not raise a TypeError.
+        unset_document = self.documents[6]
+        CustomFieldInstance.objects.create(
+            document=unset_document,
+            field=self.custom_fields["documentlink_field"],
+            value_document_ids=None,
+        )
+        self._assert_query_match_predicate(
+            ["documentlink_field", "contains", [unset_document.id]],
+            lambda document: (
+                "documentlink_field" in document
+                and document["documentlink_field"] is not None
+                and set(document["documentlink_field"]) >= {unset_document.id}
+            ),
+            match_nothing_ok=True,
+        )
+
+    def test_document_link_contains_ignores_unrelated_document_link_field(
+        self,
+    ) -> None:
+        # A document referenced in the value list may have a *different*
+        # Document Link custom field attached with no value set. This must
+        # not be pulled into the reverse-link lookup for the field being
+        # queried, and must not crash.
+        unrelated_field = CustomField.objects.create(
+            name="unrelated_documentlink_field",
+            data_type=CustomField.FieldDataType.DOCUMENTLINK,
+        )
+        # self.documents[0] is reciprocally linked from self.documents[35]
+        # (documentlink_field=[documents[0].id, documents[1].id, documents[2].id]).
+        target_document = self.documents[0]
+        CustomFieldInstance.objects.create(
+            document=target_document,
+            field=unrelated_field,
+            value_document_ids=None,
+        )
+
+        self._assert_query_match_predicate(
+            ["documentlink_field", "contains", [target_document.id]],
+            lambda document: (
+                "documentlink_field" in document
+                and document["documentlink_field"] is not None
+                and set(document["documentlink_field"]) >= {target_document.id}
+            ),
         )
 
     # ==========================================================#

